### PR TITLE
CHAIN-123: require signature on order update

### PR DESCRIPTION
**Pull request by bflood366**

### CODE REVIEW

File: backend/src/main/kotlin/co/chainring/apps/api/services/ExchangeApiService.kt [hunk 1]
```diff
@@ -311,15 +324,21 @@ class ExchangeApiService(
...
-                    EIP712Helper.computeHash(
-                        tx,
-                        blockchainClient.chainId,
-                        verifyingContract,
-                    ),
-                    tx.signature,
-                    walletAddress,
-                )
+            if (
+                try {
```
Comment (antonyshchenko) [inline backend/src/main/kotlin/co/chainring/apps/api/services/ExchangeApiService.kt:328]: Not sure how I feel about try/catch inside an if-statement. And in general I think this method has too much nesting which makes it hard to follow. Why don't we check the presence of exchange contract first and then wrap the rest into try/catch?
E.g something like this:

```
private fun validateOrderSignature(walletAddress: Address, marketId: MarketId, amount: BigInteger, side: OrderSide, price: BigDecimal?, nonce: String, signature: EvmSignature) {
  val verifyingContract = blockchainClient.getContractAddress(ContractType.Exchange) ?: throw ExchangeError("No deployed contract found")
  val isValidSignature = try {
    val (baseSymbol, quoteSymbol) = marketId.value.split("/")
    ...
    val hash = EIP712Helper.computeHash(...)
    ECHelper.isValidSignature(hash, tx.signature, walletAddress))
  } catch (e: Exception) {
    logger.warn(e) { "Exception validating signature" }
    throw ExchangeError("Invalid signature")
  }
  if (!isVaildSignature) {
    throw ExchangeError("Invalid signature")
  }
}
```

Or like this:
```
private fun validateOrderSignature(walletAddress: Address, marketId: MarketId, amount: BigInteger, side: OrderSide, price: BigDecimal?, nonce: String, signature: EvmSignature) {
  val verifyingContract = blockchainClient.getContractAddress(ContractType.Exchange) ?: throw ExchangeError("No deployed contract found")
  runCatching {
    val (baseSymbol, quoteSymbol) = marketId.value.split("/")
    ...
    val hash = EIP712Helper.computeHash(...)
    ECHelper.isValidSignature(hash, tx.signature, walletAddress))
  }.onFailure { e ->
    logger.warn(e) { "Exception validating signature" }
  }.getOrDefault(false).also { isValidSignature ->
    if (!isValidSignature) {
      throw ExchangeError("Invalid signature")
    }
  }
}
```

File: integrationtests/src/main/kotlin/co/chainring/integrationtests/utils/Wallet.kt [hunk 2]
```diff
@@ -96,6 +97,17 @@ class Wallet(
         return request.copy(signature = blockchainClient.signData(EIP712Helper.computeHash(tx, blockchainClient.chainId, exchangeContractAddress)))
     }
 
+    fun signOrder(request: UpdateOrderApiRequest.Limit): UpdateOrderApiRequest.Limit {
+        val (baseSymbol, quoteSymbol) = request.marketId.value.split("/")
+        val tx = request.toEip712Transaction(
```
Comment (antonyshchenko) [inline integrationtests/src/main/kotlin/co/chainring/integrationtests/utils/Wallet.kt:102]: Seems like this could be simpler:

In `UpdateOrderApiRequest.Limit`:

```
fun toEip712Transaction(sender: WalletEntity, baseToken: SymbolEntity, quoteToken: SymbolEntity)
```

Then here:

```
va tx = request.toEip712Transaction(this, symbols.first { it.name == baseSymbol }, symbols.first { it.name == quoteSymbol })
```

File: web-ui/src/components/Screens/HomeScreen/OrdersAndTradesWidget.tsx [hunk 3]
```diff
@@ -266,26 +275,45 @@ export default function OrdersAndTradesWidget({
...
   const quoteSymbol = market.quoteSymbol
+  const { signTypedDataAsync } = useSignTypedData()
 
-  const [amount, setAmount] = useState(
-    formatUnits(order.amount, baseSymbol.decimals)
-  )
-  const [price, setPrice] = useState(
-    order.type == 'limit' ? String(order.price) : ''
-  )
+  const {
```
Comment (antonyshchenko) [inline web-ui/src/components/Screens/HomeScreen/OrdersAndTradesWidget.tsx:298]: 👍

**Review by antonyshchenko — APPROVED** [review]
Looks great!

File: integrationtests/src/main/kotlin/co/chainring/integrationtests/utils/Wallet.kt (hunk 2 above)
Comment (bflood366) [reply integrationtests/src/main/kotlin/co/chainring/integrationtests/utils/Wallet.kt:101]: simplified per your suggestion
